sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs

```diff
@@ -0,0 +1,309 @@
+        {
+            _acrScopes = new Dictionary<string, string>();
+            _acrAccessTokens = new Dictionary<string, AcrAccessToken>();
+            _mode = LoginMode.TokenAad;
+
+            // Proofing in case passed in loginurl includes https start.
+            if (loginUrl.StartsWith("https://"))
+            {
+                loginUrl.Substring("https://".Length);
+            }
+            if (loginUrl.EndsWith("/"))
+            {
+                loginUrl.Substring(0, loginUrl.Length - 1);
+            }
+            _loginUrl = loginUrl;
+            _requestCancellationToken = cancellationToken;
+            _aadAccess = new AuthToken(aadAccessToken, acquireNewAad);
+            _acrRefresh = new AcrRefreshToken(_aadAccess, _loginUrl);
+            _tenant = tenant;
+        }
+
+        #endregion
+
+        #region Overrides
+
+        /// <summary>
+        /// Called on initialization of the credentials. This sets forth the type of authorization to be used if necessary.
+        /// </summary>
+        public override void InitializeServiceClient<AzureContainerRegistryClient>(ServiceClient<AzureContainerRegistryClient> client)
+        {
+            if (_mode == LoginMode.Basic) // Basic Authentication
+            {
+                _authHeader = EncodeTo64(_username + ":" + _password);
+            }
+        }
+
+        /// <summary>
+        /// Handles all requests of the SDK providing the required authentication along the way.
+        /// </summary>
+        public override async Task ProcessHttpRequestAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException("request");
+            }
+
+            if (_mode == LoginMode.Basic)
+            {
+                request.Headers.Authorization = new AuthenticationHeaderValue("Basic", _authHeader);
+            }
+            else
+            {
+                string operation = "https://" + _loginUrl + request.RequestUri.AbsolutePath;
+                string scope = await getScope(operation, request.Method.Method, request.RequestUri.AbsolutePath);
+
+                request.Headers.TryAddWithoutValidation("Authorization", "Bearer " + getAcrAccessToken(scope));
+            }
+            await base.ProcessHttpRequestAsync(request, cancellationToken);
+
+        }
+
+        #endregion
+
+        #region Helpers
+
+        /// <summary>
+        /// Acquires a new ACR access token if necessary. It can also acquire a cached access token in order to avoid extra requests to
+        /// the oauth2 endpoint improving efficiency.
+        /// <param name='scope'> The scope for the particuar operation. Can be obtained from the Www-Authenticate header.
+        /// </summary>
+        public string getAcrAccessToken(string scope)
```

Comment:
Public method should start with upper case letter.

Reply:
Or should this become private?

Reply:
Seems like these are helpers and should be private. I will change the access modifier and convert to pascal case.